### PR TITLE
feat: Basic CI Templates for GA and AzureDevOps for new Extensions.Template project

### PR DESCRIPTION
**Pull request by arqueror**
## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?

We don't include any template for CI out of the box with the extensions project template.

## What is the new behavior?

New CI templates are included for GA and AzureDevOps when template is used for creating a new project. Templates are able to build and publish release artifacts for all the UNO supported platforms. Templates are also able to produce signed artifacts for Windows, Android and iOS.

new CLI paramater:
`-c azure/github `

## PR Checklist
- [x] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

### CODE REVIEW

File: src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json [hunk 1]
```diff
@@ -772,6 +819,33 @@
...
+            "build": "azure"
+          },
+          "exclude": [
+            "github/**/*"
+          ]
+        },
+        {
+          "condition": "(useGithubActions)",
+          "rename": {
+            "github": ".github"
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:835]: Is there a limitation with templates where we can't just have .github as the folder name? or are we concerned that it might impact the workflows in the extensions repo?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/Directory.Build.props [hunk 2]
```diff
@@ -28,7 +28,7 @@
 	</PropertyGroup>
 
 	<PropertyGroup>
-		<RuntimeIdentifier Condition="$(IsIOS)">iossimulator-x64</RuntimeIdentifier>
+		<RuntimeIdentifier Condition="$(IsIOS)">iossimulator-arm64</RuntimeIdentifier>
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/Directory.Build.props:31]: What's the reason for this change?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.csproj [hunk 3]
```diff
@@ -143,6 +143,7 @@
 				<!-- https://github.com/xamarin/xamarin-macios/issues/14812 -->
 				<MtouchExtraArgs>$(MtouchExtraArgs) --marshal-objectivec-exceptions:disable</MtouchExtraArgs>
 				<MtouchExtraArgs>$(MtouchExtraArgs) --registrar:static</MtouchExtraArgs>
+				<RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.csproj:146]: This shouldn't be set here - it's set in the Directory.Build.props file

File: src/Uno.Extensions.Templates/content/unoapp-extensions/MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.nocpm.csproj [hunk 4]
```diff
@@ -141,6 +141,7 @@
 				<MtouchExtraArgs>$(MtouchExtraArgs) --registrar:static</MtouchExtraArgs>
 				<!-- https://github.com/xamarin/xamarin-macios/issues/14812 -->
 				<MtouchExtraArgs>$(MtouchExtraArgs) --marshal-objectivec-exceptions:disable</MtouchExtraArgs>
+				<RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.nocpm.csproj:144]: See previous comment - this is set in Directory.Build.props

File: src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/install-dependencies.yml [hunk 5]
```diff
@@ -0,0 +1,123 @@
+parameters:
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/install-dependencies.yml:1]: Please rework this to use uno-check <-- this should be our standard tool for installing dependencies and is used across nearly all existing builds

File: src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/ios.yml [hunk 6]
```diff
@@ -0,0 +1,107 @@
+    ###################################################################
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/ios.yml:1]: We should discuss this - it is possible for us to update the plist based on the certs and provisioning profile. For example check out this script: https://github.com/builttoroam/pipeline_templates/blob/b7ff7ce2ce5e41ac0702e0ceb303b8130392dfb2/azure/stages/build-xamarin-ios.yml#L188

Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/ios.yml:1]: We should discuss this - it is possible for us to update the plist based on the certs and provisioning profile. For example check out this script: https://github.com/builttoroam/pipeline_templates/blob/b7ff7ce2ce5e41ac0702e0ceb303b8130392dfb2/azure/stages/build-xamarin-ios.yml#L188

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+            azure_subscription: '${{ variables.azureAccount }}'
+            install_mobileWorkloads: 'true'
+
+        - template: /azure/steps/android.yml
+          parameters:
+            azure_subscription: '${{ variables.azureAccount }}'
+            package_id: 'Android'
+            project_to_build: 'MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.csproj'
+
+    - job: iOS
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:59]: This job should be run on a macos vmImage (same with maccatalyst and macos)

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 8]
```diff
@@ -0,0 +1,217 @@
+trigger:
+- master
+- main
+- release
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:4]: Do we have to adust this in order to be triggered on pull requests too?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  #[REQUIRED]
+  #For connecting your azureSubscription to the pipeline:
+  #1.- Go to your project settings (Bottom Left option)
+  #2.- Scroll down to 'Service Connections'
+  #3.- Click on 'Create new service connection' button
+  #4.- Click on 'Azure Resource Manager' and ckick on Next
+  #5.- Select ''Service Principal (automatic) and click on Next
+  #6.- Select the subscription you want to use and set a Name for your new service connection
+  #7.- Finally, use the name you defined in previous step as value for this variable
+- name: azureAccount
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:19]: Why is an azure account required as part of the CI pipeline, other than for publishing perhaps?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+- name: jobTimeout
+  value: '40'
+
+stages:
+  - stage: Build_UNO
+    pool:
+      vmImage: 'windows-latest'
+
+    jobs:
+    - job: Wasm
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:31]: We need to put conditional inclusion tags around each of these jobs so that the jobs are only included for the platform selected when running the template

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml (hunk 9 above)
Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:19]: Hey Nick i am not sure because at least in my case it is always asking me for a connected account for being able to execute the scripts. Not sure it it's because i'm using a free account.

File: src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json (hunk 1 above)
Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:835]: i don't remember the exact reason but i discussed this with @dansiegel

Comment (dansiegel) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:835]: @nickrandolph this was at my request. We can discuss later

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml (hunk 7 above)
Comment (dansiegel) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:59]: I would add Android to that list... it seems to build quicker & more reliably for me on MacOS than Windows hosts?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+# trigger this pipeline if there's a PR to any of these branches
+pr:
+- master
+- main
+- staging
+- releases/*
+
+parameters:
+- name: Wasm
+  type: string
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:15]: Should these be boolean values?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/wasm.yml [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+- name: minor
+  type: string
+  default: '0'
+- name: package_id
+  type: string
+  default: 'release_Wasm'
+- name: project_to_build
+  type: string
+  default: ''
+- name: project_folder
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/wasm.yml:14]: We should be able to infer project_folder from the project_to_build filename. We can leave this as a parameter in case it needs to be overridden but if the value is '' we should infer the value from the project filename

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+      pool:
+        vmImage: macOS-latest   
+
+    - job: Android
+      condition:  eq('${{ parameters.Android }}', 'true')
+      timeoutInMinutes: '${{ variables.jobTimeout }}'
+      steps:
+        - template: /azure/steps/install-dependencies.yml  
+          parameters:
+            dotnet_version: '${{ variables.dotnet_version }}'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:73]: The parameter value matches the default, so we can remove this from the azure-pipeline.yml

File: src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json [hunk 14]
```diff
@@ -363,6 +363,30 @@
         "wasm": "true"
       }
     },
+    "continuousIntegration": {
+      "displayName": "CI Templates",
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:367]: We should make CI conditional on selecting net7.0, to avoid us having to support CI pipelines for both net6 and 7
cc @dansiegel are you ok with this?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml (hunk 13 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:73]: This comment applies to all parameters where the value being passed in matches the default - let's keep this file as simple as possible

File: src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/mobile.yml [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  default: 'release_mobile'
+- name: project_to_build
+  type: string
+  default: ''
+- name: project_folder
+  type: string
+  default: ''         
+- name: target_framework
+  type: string
+  default: 'net7.0-android'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/mobile.yml:19]: This default value doesn't make much sense, set this to '' or don't specify a value to make it a required parameter

File: src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/mobile.yml [hunk 16]
```diff
@@ -0,0 +1,75 @@
+parameters:
+- name: major
+  type: string
+  default: '0'  
+- name: minor
+  type: string
+  default: '0'
+- name: package_id
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/mobile.yml:8]: This parameter shouldn't be required since it can be inferred from the target framework

File: src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json (hunk 14 above)
Comment (dansiegel) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:367]: No... it should not be conditional. The changes between the two should be fairly easy and we should be able to adopt some conditional logic in the template.json to switch values for the environment setup between net6/net7

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml (hunk 11 above)
Comment (dansiegel) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:15]: better question... why the heck are we even including these variables to conditionally build a platform... this makes no sense

File: src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json [hunk 17]
```diff
@@ -1026,6 +1065,33 @@
...
+          ]
+        },
+        {
+          "condition": "(useGithubActions)",
+          "rename": {
+            "github": ".github"
+          },
+          "exclude": [
+            "azure-pipelines.yml",
+            "build/**/*"
```
Comment (dansiegel) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:1085]: also let's change this directory back to Azure. We can do the rename like we are for github. This makes it easier to manage long term if we have other options in the future like Nuke Build that would also want a build directory.

**Review by jeromelaban — COMMENTED** [review]
I do not see anything related to package versioning, this is required to publish on testflight and the play store. Did you creating this from gitversion?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/ios.yml [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+                $filecontent = Get-Content -Path $manifest -Raw
+                $filecontent -replace "</dict>","
+                <key>CFBundleIdentifier</key>
+                <string>$bundleid</string>
+            </dict>"| Set-Content -Path $manifest
+                Get-Content $manifest -Raw
+            }
+          }
+          
+          #Replace .csproj default ApplicationId by valid BundleId if it is not in there yet
```
[the remark this answers is not in the document]
Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/build/steps/ios.yml:172]: it's in there just to make sure the bundleId is in place in Info.Plist and csproj ApplicationId tag before trying to build. If it is not in there it extracts it from the provisioning profile and sets it in both files.

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml (hunk 11 above)
Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml]: removed this on both

Comment (dansiegel) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml]: Do keep in mind @arqueror we do actually need conditions using the Template Symbols that already exist to control whether or not to include the jobs for each of the various platforms. You might need to do some testing to determine what it will accept for the conditions... I don't know if yml files are actually supported out of the box or we need to define how the template processor should handle it... I would start with:
```yml
#if (ios)
job: Build iOS
....
#endif
```

If that doesn't work ping me and I'll work with you to get something working.

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+        - template: /azure/steps/install-dependencies.yml  
+          parameters:
+            target_platform: 'android'
+        - template: /azure/steps/android.yml
+          parameters:
+            sign_package: 'true'
+            certificate:  '' #YOUR_CERT.jks stored in secureFiles
+            certificate_password: 'contoso'
+            certificate_alias: 'contoso'
+            package_id: 'signed_android'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:72]: Elevate certificate, certificate_password and certificate_alias into pipeline parameters and then make this job conditional on those properties being true.

File: src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json [hunk 20]
```diff
@@ -534,6 +553,25 @@
         ]
       }
     },
+    "ciTargetDotNetVersion": {
+      "type": "generated",
+      "generator": "switch",
+      "replaces": "$ciTargetDotNetVersion$",
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:559]: Use 6.0 and 7.0 as target dotnet version. On the off-chance that they ship a 6.1 or 7.1 it'll make it easier to update the template

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+      timeoutInMinutes: '${{ variables.jobTimeout }}'
+      steps:
+        - template: /azure/steps/install-dependencies.yml  
+          parameters:
+            target_platform: 'ios'
+
+        - template: /azure/steps/mobile.yml
+          parameters:
+            project_to_build: 'MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.csproj'
+            target_framework: 'net$ciTargetDotNetVersion$.0-ios'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:89]: Do you need to add target_framework to the other jobs that use the mobile.yml steps (eg Android, the iOS_signed)

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+- main
+- staging
+- releases/*
+
+variables:
+- name: jobTimeout
+  value: '60'
+
+stages:
+  - stage: Build_UNO
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:18]: Currently the Build_UNO stage has two steps for android/ios/windows, with the second step being the creation of the signed package. I was wondering whether we should push the second step into a separate stage that would only be done on successful completion of the first stage. This way we're not generating signed packages unless all the platforms build successfully.
cc @dansiegel @jeromelaban  thoughts?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+        - template: /azure/steps/ios.yml
+          parameters:
+            package_id: 'signed_ios'
+            project_to_build: 'MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.csproj'
+            certificate: '' #YOUR_CERT.p12 stored in secureFiles
+            certificate_password: 'contoso'
+            provisioning_profile: ''  #YOUR_PROV_PROFILE.mobileprovision stored in secureFiles
+            codesign_provision: 'MyExtensionsApp'
+            codesign_key: 'Apple Development: Contoso Contoso (572224A42L)'
+      pool:
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:116]: See previous comment on android signing job about signing parameters - we should only enable this job if those parameters are available and should show warning if they're not

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/android.yml [hunk 24]
```diff
@@ -0,0 +1,116 @@
+parameters:
+- name: major
+  type: string
+  default: '0'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/android.yml:4]: Can we have an approach to versioning so that all platforms end up with same version?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/android.yml [hunk 25]
```diff
@@ -0,0 +1,116 @@
+parameters:
+- name: major
+  type: string
+  default: '0'  
+- name: minor
+  type: string
+  default: '0'
+- name: package_id
+  type: string
+  default: 'release_Wasm'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/android.yml:10]: "release_wasm" as default sounds wrong for android.yml?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/android.yml [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+            if ('true' -eq '${{ parameters.publish_artifacts }}') {      
+              if ('true' -eq '${{ parameters.compress_artifacts }}') {
+                Write-Host "Compressing final package"
+                $finalPackageName = "$finalPackageFolder\${{ parameters.package_id }}_${{parameters.major}}.${{parameters.minor}}.$(Build.BuildNumber).zip"
+                Compress-Archive -Path "$projectFolder\bin\Release\*" -DestinationPath $finalPackageName
+              }else{
+                Copy-Item -Path "$projectFolder\bin\Release\*" -Destination "artifacts" -recurse -Force  
+              }
+            }
+        azurePowerShellVersion: 'LatestVersion'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/android.yml:108]: Is the azurePowerShellVersion parameter required for the PowerShell task?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+  default: 'release_Desktop'
+- name: project_to_build
+  type: string
+  default: ''  
+- name: project_folder
+  type: string
+  default: ''    
+- name: project_platform
+  type: string
+  default: 'x86'
```
Comment (nickrandolph) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml:19]: Should we default to x64 here? and/or generate both platforms?
cc @dansiegel @jeromelaban

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml (hunk 22 above)
Comment (jeromelaban) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:52]: Signing is generally best to be done on a separate stage to avoid rebuilding multiple times on signing failures (in case a provisioning profile needs to be provided again)

Comment (dansiegel) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:52]: It is one thing if we want to produce an unsigned artifact for a PR but when this runs for a merge then it should always produce a signed artifact otherwise what's the point of the template. The notion that you need to wait and see if all of the platforms build successfully and then go back and build it a second time to produce a signed version would be a waste of agent build minutes.

The more typical for a merge is build everything, sign what needs to be signed... then if they all build successfully you trigger a deployment stage which could be to say QA... then if that is approved you might rebuild your app using a different environment so that your app hits your prod API, etc... then deploy to prod... In any case the default should always be produce signed builds with the exception of maybe for a PR....

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml (hunk 27 above)
Comment (dansiegel) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml:13]: x64... does Apple actually make an x86 device anymore???

File: src/Uno.Extensions.Templates/content/unoapp-extensions/github/workflows/steps/build_ios/action.yml [hunk 28]
```diff
@@ -0,0 +1,228 @@
...
+            if ('' -eq '${{ inputs.certificate-password }}') { $errors = $errors +'[certificate-password] is not valid. Please make sure you are using the correct .p12 certificate password' + "`r`n"  }
+            if ('' -eq '${{ inputs.codesign-key }}') { $errors = $errors +'[codesign-key] is not valid. Please make sure the name of the name of the distribution certificate is correct. E.g. Apple Development: Contoso Contoso (752234A22L)' + "`r`n"  }
+            if ('' -eq '${{ inputs.codesign-provision }}') { $errors = $errors +'[codesign-provision] is not valid. Please make sure the provisioning profile name is correct. This is the name you entered in the Apple Developer portal when creating your provisioning profile' + "`r`n"  }
+            if ('' -eq '${{ inputs.provisioning-profile }}') { $errors = $errors +'[provisioning-profile] is not valid. Please make sure the .mobileprovision file is correct and it exists in secure files' + "`r`n"  }
+            
+            if (-not ([string]::IsNullOrEmpty($errors))){
+                $errors
+                throw "Some required parameters are missing"
+            }
```
Comment (dansiegel) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/github/workflows/steps/build_ios/action.yml:74]: scripts should be in a scripts directory where it is easier to get linting and intellisense and should not be inline like this when they are more than a couple of simple lines....

Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/github/workflows/steps/build_ios/action.yml:74]: @dansiegel  tried to include separate ps1 scripts but template did not let me to doing so because it represents a high security risk and that's why i included inline scripts instead which in theory represents the same risk but for some reason it let me do it that way ¯\_(ツ)_/¯

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/gitversion.yml [hunk 29]
```diff
@@ -0,0 +1,18 @@
+steps:
+  - task: UseDotNet@2
+    displayName: 'Use .Net Core runtime 3.1.x'
+    inputs:
+      packageType: runtime
+      version: 3.1.3
```
Comment (dansiegel) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/gitversion.yml:6]: why are we installing dotnet core 3.1?

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+        targetType: 'inline'
+        script: |
+            $projectFolder = '${{ parameters.project_folder }}'
+            if (('' -eq $projectFolder) -and ("${{ parameters.project_to_build }}".Contains('/') )){
+              Write-Host "Adjusting proj folder"
+              $projectFolder = "${{ parameters.project_to_build }}".SubString(0, "${{ parameters.project_to_build }}".LastIndexOf('/'))
+            }
+            Write-Host "Current proj folder is: $projectFolder"
+
+            Write-Host  ""
```
Comment (dansiegel) [inline src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml:51]: This all seems unnecessary... the template should show what a valid input looks like.. and that shouldn't be something that changes... we should need to clean up a bad input.

**Review by dansiegel — CHANGES_REQUESTED** [review]
This is looking better. But I'm really not a huge fan of the over validation. This makes the templates heavy and gives the feeling that this is a much harder process than it really is or should be. IMO this should be as simple as possible just `dotnet publish <arguments>` instead of crazy inline scripts that we have going on. 
We should only be falling back to insert extra validation if it's absolutely necessary. cc: @nickrandolph @jeromelaban

File: src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json [hunk 31]
```diff
@@ -1036,6 +1084,25 @@
...
+          ]
+        },
+        {
+          "condition": "(!useAzureDevOps)",
+          "rename": {
+            "github": ".github"
+          },
+          "exclude": [
+            "azure-pipelines.yml",
+            "azure/**/*"
```
[the remark this answers is not in the document]
Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/.template.config/template.json:1104]: yes it was but in another comment you asked me to change it back to azure. i can change it to build again

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 32]
```diff
@@ -0,0 +1,306 @@
...
+        - template: /azure/steps/install-dependencies.yml  
+          parameters:
+            target_platform: 'ios'
+
+        - template: /azure/steps/mobile.yml
+          parameters:
+            project_to_build: 'MyExtensionsApp.Mobile/MyExtensionsApp.Mobile.csproj'
+            target_framework: 'net$ciTargetDotNetVersion$-ios'
+      pool:
+        vmImage: windows-latest
```
[the remark this answers is not in the document]
Comment (nickrandolph) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:156]: Yeh I think some conditional logic on whether a) the signing info is available and b) what type of build is being run. eg IPA shouldn't be generated on PR build but should be done on a merge (at least on merge to master or release)

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml (hunk 30 above)
Comment (nickrandolph) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/desktop.yml:51]: Specifying project folder is unnecessary in most scenarios since it can be inferred from the csproj. We can clean this up completely by removing the parameter. Only reason I left it was just in case they want to push to a different output folder.

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/gitversion.yml (hunk 29 above)
Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure/steps/gitversion.yml:6]: this is required by gitversion

File: src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml [hunk 33]
```diff
@@ -0,0 +1,306 @@
...
+      timeoutInMinutes: '${{ variables.jobTimeout }}'
+      steps:
+        - template: /azure/steps/install-dependencies.yml  
+          parameters:
+            target_platform: 'android'
+
+        - template: /azure/steps/android.yml
+          parameters:
+            package_id: 'Android'
+            target_framework: "net$ciTargetDotNetVersion$-android"
```
[the remark this answers is not in the document]
Comment (arqueror) [reply src/Uno.Extensions.Templates/content/unoapp-extensions/azure-pipelines.yml:90]: currently only being used by install-dependencies.yml for setting dotnet version 6.0.x/7.0.x

**Comment by nickrandolph** [comment]
Hi all - pushing this to draft to make it clear that we're going to do final review and close this off after 4.8 and before 5.0

**Review by dansiegel — COMMENTED** [review]
When you rebase be sure to update all of your file paths with a find/replace `MyExtensionsApp` to `MyExtensionsApp.1`

**Comment by dansiegel** [comment]
Closing as we have removed the Template from this repo. The changes in this branch will need to be brought across for unoplatform/uno.templates#8 and a new PR opened there
